clients/postgres/channels.go

```diff
@@ -1,120 +1,183 @@
-	return channels
+	return items
 }
 
 func (cr channelRepository) Remove(owner, id string) error {
-	cr.db.Delete(&clients.Channel{}, "owner = ? AND id = ?", owner, id)
+	q := `DELETE FROM channels WHERE id = $1 AND owner = $2`
+	cr.db.Exec(q, id, owner)
 	return nil
 }
 
 func (cr channelRepository) Connect(owner, chanID, clientID string) error {
-	// This approach can be replaced by declaring composite keys on both tables
-	// (clients and channels), and then propagate them into the m2m table. For
-	// some reason GORM does not infer these kind of connections well and
-	// raises a "no unique constraint for referenced table". Until we find a
-	// way to properly represent this relationship, let's stick with the nested
-	// query approach and observe its behaviour.
-	sql := `INSERT INTO channel_clients (channel_id, client_id)
-	SELECT ?, ? WHERE
-	EXISTS (SELECT 1 FROM channels WHERE owner = ? AND id = ?) AND
-	EXISTS (SELECT 1 FROM clients WHERE owner = ? AND id = ?);`
-
-	res := cr.db.Exec(sql, chanID, clientID, owner, chanID, owner, clientID)
-
-	if res.Error == nil && res.RowsAffected == 0 {
-		return clients.ErrNotFound
+	q := `INSERT INTO connections (channel_id, channel_owner, client_id, client_owner) VALUES ($1, $2, $3, $2)`
+
+	_, err := cr.db.Exec(q, chanID, owner, clientID)
+	if err != nil {
+		pqErr, ok := err.(*pq.Error)
+
+		if ok && errFK == pqErr.Code.Name() {
+			return clients.ErrNotFound
+		}
+
+		// connect is idempotent
+		if ok && errDuplicate == pqErr.Code.Name() {
+			return nil
+		}
+
+		return err
 	}
 
-	return res.Error
+	return nil
 }
 
 func (cr channelRepository) Disconnect(owner, chanID, clientID string) error {
-	// The same remark given in Connect applies here.
-	sql := `DELETE FROM channel_clients WHERE
-	channel_id = ? AND client_id = ? AND
-	EXISTS (SELECT 1 FROM channels WHERE owner = ? AND id = ?) AND
-	EXISTS (SELECT 1 FROM clients WHERE owner = ? AND id = ?);`
+	q := `DELETE FROM connections
+	WHERE channel_id = $1 AND channel_owner = $2
+	AND client_id = $3 AND client_owner = $2`
 
-	res := cr.db.Exec(sql, chanID, clientID, owner, chanID, owner, clientID)
+	res, err := cr.db.Exec(q, chanID, owner, clientID)
+	if err != nil {
+		return err
+	}
+
+	cnt, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
 
-	if res.Error == nil && res.RowsAffected == 0 {
+	if cnt == 0 {
 		return clients.ErrNotFound
 	}
 
-	return res.Error
+	return nil
 }
 
 func (cr channelRepository) HasClient(chanID, clientID string) bool {
-	sql := "SELECT EXISTS (SELECT 1 FROM channel_clients WHERE channel_id = $1 AND client_id = $2);"
-
-	row := cr.db.DB().QueryRow(sql, chanID, clientID)
-
-	var exists bool
-	if err := row.Scan(&exists); err != nil {
-		// TODO: this error should be logged
-		return false
-	}
+	q := "SELECT EXISTS (SELECT 1 FROM connections WHERE channel_id = $1 AND client_id = $2);"
 
+	// TODO: log errors?
+	exists := false
+	cr.db.QueryRow(q, chanID, clientID).Scan(&exists)
```

Comment:
I think that this error should be logged. There is not other way for us to know that something went wrong.